### DISCUSSION TITLE
Better deadman switch for keyboard control?

### DISCUSSION THREAD

**Discussion by falkoschindler** [zauberzeug/rosys#173] [Ideas / Feature Requests]
The `keyboard_control` in RoSys currently regularly asks the client if it's still connected. If the response comes too late or not at all, the robot automatically stops.

However, due to the JavaScript command, the `keyboard_control` cannot be used on the auto-index client, which means some examples are currently not running without errors. That's why we had been thinking about implementing the deadman switch differently.

How about if we provide an optional ping in NiceGUI's `ui.keyboard` that is sent from the client at a specific interval and potentially triggers a timeout, which can then call a timeout handler? This isn't 100% secure on the auto-index page, as a ping from one browser would be received while the connection to another might be dropping. But it's relatively simple and allows for keyboard control in basic examples.

Strictly speaking, such a deadman feature would be relatively independent of the keyboard and could also be its own UI element. So maybe we want a `ui.ping` or `ui.deadman_switch` that we can integrate into the keyboard control. Would this be a NiceGUI or a RoSys element?

**Comment by rodja** [comment]
Cool idea. Another name would be `ui.connection_guard` or `ui.heartbeat`. It should take two parameters: a callback and the seconds of non-connectivity until callback is executed. This would be cool to have in NiceGUI. There might be more use cases than robotics: real time collaborative sites, multi-player games, home automation, ...
